### PR TITLE
rgblight.h no longer changes behavior when included >1 time

### PR DESCRIPTION
**Pull request by GrayHatter**
## Description

Headers should never change behavior dependent on the number of times
they're included. This fixes that defect.

## Types of Changes

- [x] Core
- [x] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

It works for me, I can't build the whole repo for obvious reasons `make test[s]` seemed to work **shrug**

### CODE REVIEW

File: quantum/rgblight.h [hunk 1]
```diff
@@ -64,19 +62,10 @@
...
-#define _RGBM_MULTI_DYNAMIC(sym)   RGBLIGHT_MODE_ ## sym,
-#define _RGBM_TMP_STATIC(sym)      RGBLIGHT_MODE_ ## sym,
-#define _RGBM_TMP_DYNAMIC(sym)     RGBLIGHT_MODE_ ## sym,
-enum RGBLIGHT_EFFECT_MODE {
-    RGBLIGHT_MODE_zero = 0,
-#include "rgblight.h"
-    RGBLIGHT_MODE_last
-};
 
-#ifndef RGBLIGHT_H_DUMMY_DEFINE
```
Comment (drashna) [inline quantum/rgblight.h]: This needs to be reverted.

This is actually required for the rgb matrix code.  If this is removed, they failed to compile.

File: quantum/rgblight.h [hunk 2]
```diff
@@ -64,19 +62,10 @@
...
+/* contains the `RGBLIGHT_EFFECT_MODE enum` */
+#include "rgblight_modes.h"
+
```
Comment (drashna) [inline quantum/rgblight.h]:
```suggestion
#ifndef RGBLIGHT_H_DUMMY_DEFINE	
```

File: quantum/rgblight.h [hunk 3]
```diff
@@ -244,73 +235,4 @@ void rgblight_effect_christmas(void);
 void rgblight_effect_rgbtest(void);
 void rgblight_effect_alternating(void);
 
-#endif // #ifndef RGBLIGHT_H_DUMMY_DEFINE
 #endif // RGBLIGHT_H
```
Comment (drashna) [inline quantum/rgblight.h]: This needs to be here. 
```suggestion
#endif // #ifndef RGBLIGHT_H_DUMMY_DEFINE	
#endif // RGBLIGHT_H
```

File: quantum/rgblight.h [hunk 4]
```diff
@@ -64,19 +62,10 @@
...
-#include "rgblight.h"
-    RGBLIGHT_MODE_last
-};
 
-#ifndef RGBLIGHT_H_DUMMY_DEFINE
+/* contains the `RGBLIGHT_EFFECT_MODE enum` */
+#include "rgblight_modes.h"
+
 
 #define RGBLIGHT_MODES (RGBLIGHT_MODE_last-1)
```
Comment (drashna) [inline quantum/rgblight.h]: This line should be in the `rgblight_modes.h`

**Comment by drashna** [comment]
Honestly, I hate this implementation, and it's .... far from clean. 
This is just moving around where it is at, and doesn't remove any tech debt. 

I'd really rather see this changed to how the rgb matrix handles things, as that is cleaner, IMO.

**Comment by GrayHatter** [comment]
There's a lot of things wrong with `rgblight.h`, and I didn't want to make a huge time investment/pull before I knew why there's so much magic. Either way, `RGBLIGHT_H_DUMMY_DEFINE` is the wrong way to go about it, so if that's needed, I think it's better to not touch any of it. 

Why does rgb_matrix depend upon, or conflict with, rgblight? and does it need to?  Perhaps the solution is better separation between the two systems? I don't use/know anything about rgb_matrix, so I don't know what the best way to go about is; but depending on consensus, I'm happy to fix up this PR.

**Comment by drashna** [comment]
because a lot of the rgb matrix functions are mapped to the rgblight functions.  
https://github.com/qmk/qmk_firmware/blob/master/quantum/rgb_matrix.h#L185-L207

This is done because the RGB Matrix feature shares the RGB Light Keycodes (rather than adding separate keycodes). 

https://github.com/qmk/qmk_firmware/blob/master/docs/feature_rgb_matrix.md#keycodes

Unfortunately, this has made a bit of a mess of things.

As for the enum, that was added by mtei, IIRC, so that certain modes could be turned off, to save space in the firmware. 


But yes, the rgblight code needs a lot of cleanup, to be honest.  
But as is, we cannot merge your initial changes, since it will outright break the rgb matrix feature. And we can't remove the dummy define, since that will likely break both features.

**Comment by GrayHatter** [comment]
Yeah, I would never expect you to merge something that breaks anything. Next time I have some more time, I'll dig into what parts are shared by both systems, and make them external to both. Then including one wont require the other.

Why does rgb_matrix add rgblight as an alias? Shouldn't it NOT do that?

Do you want to close this PR, and I'll open a new once I get the above issues fixed?

**Comment by drashna** [comment]
>Why does rgb_matrix add rgblight as an alias? Shouldn't it NOT do that?

I think mostly because of the rgb keycodes.    But I think @jackhumbert is the one that would know? 

>Do you want to close this PR, and I'll open a new once I get the above issues fixed?

You can keep this one open and work on it in the meanwhile.  I'll just mark it as "in progress" until it's ready.

**Comment by drashna** [comment]
Just brainstorming here, to remove the rgblight functions, we'd need to add the rgb_matrix functions explicitly to the RGB_* keycodes in quantum.c.  

Additionally, we'd probably want to add some sort of define so that we can target rgblight or rgbmatrix, in case you want control of one or the other. 

Eg, I use rgblight for layer indication primarily.  So, I'd want the keycodes to target rgb matrix, rather than rgblight.  But others may want the opposite. 

Additionally, if you want both, you absolutely need to set: https://docs.qmk.fm/#/feature_rgb_matrix?id=eeprom-storage

**Comment by drashna** [comment]
There are a number of merge conflicts here, now.

Also, I think that this behavior may have been changed in #5020.

**Comment by GrayHatter** [comment]
I was told on discord this was fixed in a different pr. I just forgot to close this. Sorry for letting this go stale

**Comment by drashna** [comment]
Well, that's on us, too. So no worries!

And yeah, closing this then.

**Pull request closed without merging**